File: packages/api/src/endpoints/datum.ts
import type { DatumDetailResponse } from "@/types/datumTypes";

import { handleFetch } from "@/lib/handleFetch";

interface DatumDetailParams {
  hash?: string;
}

/**
 * Fetches the detail of a datum by its hash.
 *
 * This function queries the `/datum/detail` endpoint with the provided hash parameter
 * to retrieve detailed information about the datum.
 *
 * @param {DatumDetailParams} params – Object containing the datum hash.
 * @param {string} params.hash – The hash of the datum to retrieve.
 *
 * @returns {Promise<DatumDetailResponse>} A promise that resolves to the datum detail response.
 *
 * @throws Will throw an error if the network request fails or the response is invalid.
 *
 * @example
 * ```ts
 * const response = await getDatumDetail({ hash: "abcd1234" });
 * console.log(response);
 * ```
 */
export const getDatumDetail = async ({ hash }: DatumDetailParams) => {
  const url = `/datum/detail`;
  const options = {
    params: { hash },
  };

  return handleFetch<DatumDetailResponse>(url, undefined, options);
};

File: packages/cexplorer-api-rs/src/endpoints/datum.rs
use crate::client::fetch_with_params;
use crate::error::CexplorerError;
use crate::types::datum_types::*;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct DatumDetailParams {
    pub hash: String,
}

pub async fn get_datum_detail(hash: &str) -> Result<DatumDetailResponse, CexplorerError> {
    let endpoint = "/datum/detail";
    let params = DatumDetailParams {
        hash: hash.to_string(),
    };
    fetch_with_params::<DatumDetailResponse, DatumDetailParams>(endpoint, Some(&params)).await
}
